### PR TITLE
Support the debugging session via serial protocol

### PR DESCRIPTION
**Pull request by robertsipka**
The communication part and the debugger functions are work fine both on Linux and Windows.
Configure parameters for serial port is also available.

### CODE REVIEW

File: src/IotjsDebugger.ts [hunk 1]
```diff
@@ -194,19 +192,38 @@ class IotjsDebugSession extends DebugSession {
...
+    }
+
     this._protocolhandler.debuggerClient = this._debuggerClient;
 
     this._debuggerClient.connect()
     .then(() => {
-      this.log(`Connected to: ${args.address}:${args.port}`, LOG_LEVEL.SESSION);
+      if (args.channel === 'websocket') {
+        this.log(`Connected to: ${args.address}:${args.port}`, LOG_LEVEL.SESSION);
+
```
Comment (rtakacs) [inline src/IotjsDebugger.ts]: Unnecessary newline?

File: src/IotjsDebuggerInterfaces.d.ts [hunk 2]
```diff
@@ -21,6 +21,10 @@ export interface IAttachRequestArguments extends DebugProtocol.AttachRequestArgu
   address: string;
   /** Debug port to attach to. */
   port: number;
+  /** Specify the communication channel */
```
Comment (rtakacs) [inline src/IotjsDebuggerInterfaces.d.ts]: There is no dot at the end of the sentence.

File: src/JerryDebuggerSerialClient.ts [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+          openOptions.dataBits = 6;
+          break;
+        case '7':
+          openOptions.dataBits = 7;
+          break;
+        case '8':
+          openOptions.dataBits = 8;
+          break;
+        default:
+          reject(new Error(`Invalid data bit (${config[2]}) for serial configuration`));
```
Comment (rtakacs) [inline src/JerryDebuggerSerialClient.ts]: Is this a sentence? (There is no dot ot exclamation mark :)) Please fix other sentences in this patch.

File: src/JerryProtocolHandler.ts [hunk 4]
```diff
@@ -1055,6 +1057,12 @@ export class JerryDebugProtocolHandler {
   }
 
   private submitRequest(request: PendingRequest, simple: boolean = false): boolean {
+    if (this.debuggerClient.channel === 'serial') {
+      let arrayBufferConcat = require('arraybuffer-concat');
+      request.data = arrayBufferConcat(encodeMessage(this.byteConfig, 'B', [request.data.byteLength]), request.data);
+
```
Comment (rtakacs) [inline src/JerryProtocolHandler.ts]: Unnecessary newline?

File: src/JerryDebuggerSerialClient.ts [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+    let portsList = [];
+
+    await SerialPort.list().then(
+      ports => ports.forEach((port) => {
+        portsList.push(port.comName);
+      })
+    );
+
+    const config = this.serialConfig.split(',', 5);
```
Comment (rtakacs) [inline src/JerryDebuggerSerialClient.ts:59]: I'd add a comment here: // Format: port,baud,databits,parity,stopbits

File: src/JerryDebuggerSerialClient.ts [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+  private onMessage(data: ArrayBuffer): void {
+    let arrayBufferConcat = require('arraybuffer-concat');
+    this.buffer = arrayBufferConcat(this.buffer, data);
+
+    let size = new Uint8Array(this.buffer)[0];
+
+    while (size + 1 <= this.buffer.byteLength) {
+      const msg = this.buffer.slice(1, size + 1);
+      this.delegate.onMessage(new Uint8Array(msg));
+      this.buffer = this.buffer.slice(size + 1, );
```
Comment (rtakacs) [inline src/JerryDebuggerSerialClient.ts]: Does the comma necessary?

**Comment by robertsipka** [comment]
@rtakacs : Thanks for the review, I've updated it accordingly.

**Review by tdusnoki — APPROVED** [review]
LGTM (informal)

**Comment by robertsipka** [comment]
Travis does not work currently. Please do not merge until it finished successfully.

**Comment by daeyeon** [comment]
Let us know freely if you need any privilege of travis. In my glance, the configurations of npm and travis seem fine. No clue about the configuration I can find from the failure log.

**Comment by robertsipka** [comment]
Thanks, I am not sure what is wrong in here. It seems that this patch is working fine in my forked repository, the Travis CI was ran successfully on every jobs with that. The logs are exactly the same until the error messages as well.

**Comment by robertsipka** [comment]
Travis CI is green on master now, but for some reason it will  not start in case of new pull requests.

**Comment by yichoi** [comment]
Seems to be fixed. please push again to trigger Travis

**Comment by robertsipka** [comment]
Yes, it seems that it works properly now, thanks.

**Review by rtakacs — APPROVED** [review]
LGTM

**Review by yichoi — APPROVED** [review]
LGTM

**Pull request merged**